kubevirt/kubevirt-tutorial
# Added Tests to Kubevirt-Tutorials

- Added testing over actual MD files with Literal Programming ([mdsh](https://github.com/bashup/mdsh))
- Created the entrypoint for Build and Testing over a Makefile
  - Build will generate the scripts from Markdown files and put all of them on `build/` folder
  - Tests will:
    - Create a K8s cluster on a DIND job over Prow platform
    - Deploy the laboratory using our Ansible playbooks (in order to test also the playbooks)
    - Execute all files (from build step) which will emulate a student executing the labs in a row
    - The test execution will report the results in 2 ways, the first one will cover the whole script file (1 lab) and is something like:
    ```
    -------
    Test Name: lab006
    Test Result: PASS
    Test Expected Result: PASS
    Real Time: 34.008
    ```
    The second one is a log execution of the lab itself which will contain this format (for every command executed in the lab):
    ```
    Command: while ! kubectl get deployment cdi-operator -n cdi --no-headers && echo 'Looping...'; do sleep 5; done
    Output:
    cdi-operator   1/1   1     1     66s
    Time: 0.247
    ```
    - At the end of the tests execution you will get a sumarization of all the tests:
    ```
    -------
    Tests Done: 11
    Tests on Passed state: 11
    Tests on Error state: 0
    ```
    - If there are an error on a lab, the Prow job will finish on Error

- Lab tests logs also are uploaded to GCS in order to store the results
  - *${ARTIFACTS}/logs*
- Some labs has been modified in order to be well tested
- `test_lab` script has been created in order to be modular, `test_lab` will be summoned from `make tests`
- Lab MD files has been renamed from *lab9.md* to *lab009.md* for scripting purposes
- Removed some *\`\`\`console* headers in order to follow *mdsh* way of work
- Retry policy (just one time) in order to avoid platform errors.
- Real time over all the commands executed and also summarized to allow the admins to find bottlenecks or errors on Lab execution.

This is an sample execution over Prow platform:
```
...
...
make[1]: Leaving directory '/workspace/kubevirt-tutorial'
-------
Test Name: lab000
Test Result: PASS
Test Expected Result: PASS
Real Time: 0
-------
Test Name: lab001
Test Result: PASS
Test Expected Result: PASS
Real Time: .517
-------
Test Name: lab002
Test Result: PASS
Test Expected Result: PASS
Real Time: .781
-------
Test Name: lab003
Test Result: ERR
Test Expected Result: PASS
Real Time: 221.473
Retrying Test...
-------
Test Name: lab003
Test Result: PASS
Test Expected Result: PASS
Real Time: 9.951
-------
Test Name: lab004
Test Result: PASS
Test Expected Result: PASS
Real Time: 94.851
-------
Test Name: lab005
Test Result: ERR
Test Expected Result: PASS
Real Time: 9.028
Retrying Test...
-------
Test Name: lab005
Test Result: PASS
Test Expected Result: PASS
Real Time: 2.589
-------
Test Name: lab006
Test Result: PASS
Test Expected Result: PASS
Real Time: 34.008
-------
Test Name: lab007
Test Result: PASS
Test Expected Result: PASS
Real Time: 0
-------
Test Name: lab008
Test Result: PASS
Test Expected Result: PASS
Real Time: 2.223
-------
Test Name: lab009
Test Result: PASS
Test Expected Result: PASS
Real Time: .258
-------
Test Name: lab010
Test Result: PASS
Test Expected Result: PASS
Real Time: 0
-------
Tests Done: 11
Tests on Passed state: 11
Tests on Error state: 0
+ EXIT_VALUE=0
+ set +o xtrace
Cleaning up after docker in docker.
================================================================================
[Barnacle] 2019/07/11 10:08:41 Cleaning up Docker data root...
[Barnacle] 2019/07/11 10:08:41 Removing all containers.
```
